src/datapool/datapool_pmem.c

```diff
@@ -0,0 +1,217 @@
+ * opened.
+ */
+struct datapool_header {
+    uint8_t signature[DATAPOOL_SIGNATURE_LEN];
+    uint64_t version;
+    uint64_t size;
+    uint64_t flags;
+    uint8_t unused[DATAPOOL_HEADER_LEN - 32];
+};
+
+struct datapool {
+    void *addr;
+
+    struct datapool_header *hdr;
+    void *user_addr;
+    size_t mapped_len;
+    int is_pmem;
+    int file_backed;
+};
+
+static void
+datapool_sync_hdr(struct datapool *pool)
+{
+    pmem_msync(pool->hdr, DATAPOOL_HEADER_LEN);
+}
+
+static void
+datapool_sync(struct datapool *pool)
+{
+    pmem_msync(pool->addr, pool->mapped_len);
+}
+
+static int
+datapool_valid(struct datapool *pool)
+{
+    if (memcmp(pool->hdr->signature,
+          DATAPOOL_SIGNATURE, DATAPOOL_SIGNATURE_LEN) != 0) {
+        log_info("no signature found in datapool");
+        return 0;
+    }
+
+    if (pool->hdr->version != DATAPOOL_VERSION) {
+        log_info("incompatible datapool version (is: %d, expecting: %d)",
+            pool->hdr->version, DATAPOOL_SIGNATURE);
+        return 0;
+    }
+
+    if (pool->hdr->size == 0) {
+        log_error("datapool has 0 size");
+        return 0;
+    }
+
+    if (pool->hdr->size > pool->mapped_len) {
+        log_error("datapool has invalid size (is: %d, expecting: %d)",
+            pool->mapped_len, pool->hdr->size);
+        return 0;
+    }
+
+    if (pool->hdr->flags & ~DATAPOOL_VALID_FLAGS) {
+        log_error("datapool has invalid flags set");
+        return 0;
+    }
+
+    if (pool->hdr->flags & DATAPOOL_FLAG_DIRTY) {
+        log_info("datapool has a valid header but is dirty");
+        return 0;
+    }
+
+    return 1;
+}
+
+static void
+datapool_initialize(struct datapool *pool)
+{
+    log_info("initializing fresh datapool");
+
+    /* 1. clear the header from any leftovers */
+    memset(pool->hdr, 0, DATAPOOL_HEADER_LEN);
+    datapool_sync_hdr(pool);
+
+    /* 2. fill in the data */
+    pool->hdr->version = DATAPOOL_VERSION;
+    pool->hdr->size = pool->mapped_len;
+    pool->hdr->flags = 0;
+    datapool_sync_hdr(pool);
+
+    /* 3. set the signature */
+    memcpy(pool->hdr->signature, DATAPOOL_SIGNATURE, DATAPOOL_SIGNATURE_LEN);
+    datapool_sync_hdr(pool);
+}
+
+static void
+datapool_flag_set(struct datapool *pool, int flag)
+{
+    pool->hdr->flags |= flag;
+    datapool_sync_hdr(pool);
+}
+
+static void
+datapool_flag_clear(struct datapool *pool, int flag)
+{
+    pool->hdr->flags &= ~flag;
+    datapool_sync_hdr(pool);
+}
+
+/*
+ * Opens, and if necessary initializes, a datapool that resides in the given
+ * file. If no file is provided, the pool is allocated through cc_zalloc.
+ *
+ * The the datapool to retain its contents, the datapool_close() call must
+ * finish successfully.
+ */
+struct datapool *
+datapool_open(const char *path, size_t size, int *fresh)
+{
+    struct datapool *pool = cc_alloc(sizeof(*pool));
+    if (pool == NULL) {
+        log_error("unable to create allocate memory for pmem mapping");
+        goto err_alloc;
```

Comment:
wonder if using `goto` really makes sense here; each `goto` label only has one `goto` sending it there

Reply:
This is a common error handling idiom when cleanup is needed before exiting the function. There should be only one `goto` per label because this allows for stacking the cleanup functions. You could say that this is poor's man RAII.

Reply:
cool, i'm fine with leaving this as is. it isn't explicitly out of line with our style guide, and i don't think it really has a negative impact on readability; i just haven't seen this idiom before.